Sources/AsyncAlgorithms/AsyncZipLatestFromSequence.swift

```diff
@@ -0,0 +1,382 @@
+  /// await other2.send("d")
+  ///
+  /// ... later in the application flow
+  ///
+  /// await base.send(1)
+  ///
+  /// // will print: (1, "b", "d")
+  /// ```
+  ///
+  /// - Parameters:
+  ///   - other1: the first other ``AsyncSequence``
+  ///   - other2: the second other ``AsyncSequence``
+  /// - Returns: an ``AsyncZipLatestFrom2Sequence`` where elements are a tuple of an element from `self` and the
+  /// latest known elements (if any) from the `other` sequences.
+  @inlinable
+  func zipLatest<Other1: AsyncSequence, Other2: AsyncSequence>(
+    from other1: Other1,
+    _ other2: Other2
+  ) -> AsyncZipLatestFrom2Sequence<Self, Other1, Other2> {
+    AsyncZipLatestFrom2Sequence(self, other1, other2)
+  }
+}
+
+/// ``AsyncZipLatestFromSequence`` is an ``AsyncSequence`` where elements are a tuple of an element from `base` and the
+/// latest known element (if any) from the `other` sequence.
+public struct AsyncZipLatestFromSequence<Base: AsyncSequence, Other: AsyncSequence>: AsyncSequence where Other: Sendable {
+  public typealias Element = (Base.Element, Other.Element)
+  public typealias AsyncIterator = Iterator
+
+  @usableFromInline
+  let base: Base
+
+  @usableFromInline
+  let other: AsyncCurrentElementSequence<Other>
+
+  /// Creates an ``AsyncSequence`` where elements are a tuple of an element from `base` and the
+  /// latest known element (if any) from the `other` sequence.
+  /// - Parameters:
+  ///   - base: the ``AsyncSequence`` that drives the production of the output tuples
+  ///   - other: the ``AsyncSequence`` from which the latest known element is gathered to form the output tuple
+  @inlinable
+  public init(_ base: Base, _ other: Other) {
+    self.base = base
+    self.other = AsyncCurrentElementSequence(other)
+  }
+
+  @inlinable
+  public func makeAsyncIterator() -> Iterator {
+    Iterator(
+      self.base.makeAsyncIterator(),
+      self.other.makeAsyncIterator()
+    )
+  }
+
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var base: Base.AsyncIterator
+
+    @usableFromInline
+    var other: AsyncCurrentElementSequence<Other>.AsyncIterator
+
+    @usableFromInline
+    var isFinished = false
+
+    @usableFromInline
+    init(_ base: Base.AsyncIterator, _ other: AsyncCurrentElementSequence<Other>.AsyncIterator) {
+      self.base = base
+      self.other = other
+    }
+
+    @inlinable
+    public mutating func next() async rethrows -> Element? {
+      guard !Task.isCancelled else {
+        self.other.task.cancel()
+        return nil
+      }
+      guard !self.isFinished else { return nil }
+
+      var baseElement: Base.Element?
+      var lastKnownOtherState: AsyncCurrentElementSequence<Other>.State?
+
+      var hasElement = false
+      repeat {
+        do {
+          baseElement = try await self.base.next()
+        } catch {
+          self.other.task.cancel()
+          throw error
+        }
+        lastKnownOtherState = await self.other.next()
```

Comment:
so this does not run the two sequences concurrently for iteration. effectively it is blocking iteration from one to the other. Perhaps a TaskGroup is in order here (like how zip does it)?

Reply:
The `Other` sequence is being iterated over in its own Task. The call to `self.other.next()` will return a value right away (either noElement, element(value) or nil). With the way it is implemented right now I don't see the benefit of having a TaskGroup.

However, the current implementation might not be the good one! What do you think of iterating over `Other` in a separate Task so that we can always have its latest known value at our disposal ? It comes with the drawback that the latest known value might not be the one that we think because the Task's execution is scheduled by the system and it might be paused for some reason.